File: src/main/scala/xieyuheng/cicada/prelude/list.scala
package xieyuheng.cicada.prelude

import xieyuheng.cicada._
import xieyuheng.cicada.expDSL._

object list extends Module {

  import_all(nat)

  define_type("list_t", %("A" -> the_type),
    members = %(
      "null_t" -> %(),
      "cons_t" -> %(
        "head" -> the("A"),
        "tail" -> the("list_t" ap %("A" -> "A")))))

  define_fn("cdr",
    args = %("list" -> the("list_t")),
    ret = the("list_t"),
    body = "list" dot "tail")

  define_fn("list_length",
    args = %("list" -> the("list_t")),
    ret = the("nat_t"),
    body = choice("list", %(
      "null_t" -> "zero_t",
      "cons_t" -> ("succ_t" ap %(
        "prev" -> ("list_length" ap %(
          "list" -> ("list" dot "tail"))))))))

  define_fn("list_append",
    args = %(
      "ante" -> the("list_t"),
      "succ" -> the("list_t")),
    ret = the("list_t"),
    body = choice("ante", %(
      "null_t" -> "succ",
      "cons_t" -> ("cons_t" ap %(
        "A" -> ("ante" dot "A"),
        "head" -> ("ante" dot "head"),
        "tail" -> ("list_append" ap %(
          "ante" -> ("ante" dot "tail"),
          "succ" -> "succ")))))))

}

File: lib/collection/list.cs
module collection.list

type list_t(A: type_t) {
  null_t
  cons_t(head: A, tail: list_t(A))
}

list_length(list: list_t): nat_t {
  list choice {
    null_t => zero_t
    cons_t => succ_t(list_length(list.tail))
  }
}

list_append(A: type_t, ante: list_t(A), succ: list_t(A)): list_t(A) {
  ante choice {
    null_t => succ
    cons_t => cons_t(ante.head, list_append(ante.tail, succ))
  }
}

list_map(A: type_t, B: type_t, f: A -> B, list: list_t(A)): list_t(B) {
  list choice {
    null_t => list
    cons_t => cons_t(fun(list.head), list_map(fun, list.tail))
  }
}
